### PR TITLE
Very inefficient decoding of large chunked messages!

### PR DESCRIPTION
**Pull request by domoran**
In one of our legacy projects we download a ~200MB file in memory using chunked encoding. Calling getBody() on this (which calls decodeChunkedBody()) will take > 5 minutes for processing the response() since creates very inefficient copies of the body with each chunk (substr). 

To make this more efficient we can use the offset parameter, to avoid copying the body during iteration. 

Tests still passed.

### CODE REVIEW

**Comment by Ocramius** [comment]
Requires a performance test to prevent regressions. Unsure how to achieve that besides using relative timing of the regex applied to some data, and a large chunk of text passed to this method.

**Comment by domoran** [comment]
I added a performance test, by crafting a "worst case" packet and measuring relative timing. 

Verified that test fails on old version: 
git checkout e89dac73d56b18023b8e2e41074e64bb3abbdfda src/Response.php

File: src/Response.php [hunk 1]
```diff
@@ -495,17 +495,25 @@ protected function decodeChunkedBody($body)
...
-        while (trim($body)) {
-            if (! preg_match("/^([\da-fA-F]+)[^\r\n]*\r\n/sm", $body, $m)) {
-                throw new Exception\RuntimeException(
-                    "Error parsing body - doesn't seem to be a chunked message"
-                );
+        $offset = 0;
+
+        while (true) {
+            if (! preg_match("/^([\da-fA-F]+)[^\r\n]*\r\n/sm", $body, $m, 0, $offset)) {
+                if (! empty(trim(substr($body, $offset)))) {
```
Comment (webimpress) [inline src/Response.php]: `! empty` seems to be useless here

File: src/Response.php [hunk 2]
```diff
@@ -495,17 +495,25 @@ protected function decodeChunkedBody($body)
...
+                    // Message was not consumed completely!
+                    throw new Exception\RuntimeException(
+                        "Error parsing body - doesn't seem to be a chunked message"
```
Comment (webimpress) [inline src/Response.php]: Can we use single quotes around the error message string above?

File: src/Response.php [hunk 3]
```diff
@@ -495,17 +495,25 @@ protected function decodeChunkedBody($body)
...
+                    );
+                } else {
```
Comment (webimpress) [inline src/Response.php]: I think we don't need `else` statement here.

File: test/ResponseTest.php [hunk 4]
```diff
@@ -176,6 +177,54 @@ public function testChunkedResponseCaseInsensitiveZF5438()
         $this->assertEquals('c0cc9d44790fa2a58078059bab1902a9', md5($res->getContent()));
     }
 
+    /**
+     * @param number $chunksize the data size of the chunk to create
```
Comment (webimpress) [inline test/ResponseTest.php]: Please use `int` param type instead of `number`

File: test/ResponseTest.php [hunk 5]
```diff
@@ -176,6 +177,54 @@ public function testChunkedResponseCaseInsensitiveZF5438()
...
+     */
+    private function makeChunk($chunksize)
+    {
+        $chunkdata = str_repeat("W", $chunksize);
+        return "$chunksize\r\n$chunkdata\r\n";
+    }
+
+    /**
+     * @param Response $response
+     * @return the time that calling the getBody function took on the response
```
Comment (webimpress) [inline test/ResponseTest.php]: Please add return type.

File: test/ResponseTest.php [hunk 6]
```diff
@@ -176,6 +177,54 @@ public function testChunkedResponseCaseInsensitiveZF5438()
         $this->assertEquals('c0cc9d44790fa2a58078059bab1902a9', md5($res->getContent()));
     }
 
+    /**
+     * @param number $chunksize the data size of the chunk to create
+     * @return string a chunk of data for embedding inside a chunked response
+     */
+    private function makeChunk($chunksize)
+    {
+        $chunkdata = str_repeat("W", $chunksize);
```
Comment (webimpress) [inline test/ResponseTest.php]: Please use single quotes `'` above instead of double `"`.

File: test/ResponseTest.php [hunk 7]
```diff
@@ -176,6 +177,54 @@ public function testChunkedResponseCaseInsensitiveZF5438()
...
+        return "$chunksize\r\n$chunkdata\r\n";
```
Comment (webimpress) [inline test/ResponseTest.php]: Not sure here, but maybe `sprintf` and/or `PHP_EOL`?

File: test/ResponseTest.php [hunk 8]
```diff
@@ -176,6 +177,54 @@ public function testChunkedResponseCaseInsensitiveZF5438()
...
+    }
+
+    /**
+     * @param Response $response
+     * @return the time that calling the getBody function took on the response
+     */
+    private function getTimeForGetBody(Response $response)
+    {
+        $time_start = microtime(true);
```
Comment (webimpress) [inline test/ResponseTest.php]: Please use camelCase variable names.

File: test/ResponseTest.php [hunk 9]
```diff
@@ -176,6 +177,54 @@ public function testChunkedResponseCaseInsensitiveZF5438()
...
+        $response->setContent($responseData);
+        $time1 = $this->getTimeForGetBody($response);
+
+        // Get baseline for timing: 1000 x 1 Byte chunks
+        $responseData2 = $responseData . $this->makeChunk(1000000);
+        $response->setContent($responseData2);
+        $time2 = $this->getTimeForGetBody($response);
+
+        // Make sure that the worst case packet will have an equal timing as the baseline
+        $errMsg = "Chunked response is not parsing large packets efficiently: " . ($time2 / $time1);
```
Comment (webimpress) [inline test/ResponseTest.php]: Please use `sprintf` there and single quotes.

File: test/ResponseTest.php [hunk 10]
```diff
@@ -176,6 +177,54 @@ public function testChunkedResponseCaseInsensitiveZF5438()
...
+        $this->assertTrue(2 > ($time2 / $time1), $errMsg);
```
Comment (webimpress) [inline test/ResponseTest.php]: Maybe we can use here `assertLessThan`?
Brackets around `$time2 / $time1` are useless.

File: test/ResponseTest.php [hunk 11]
```diff
@@ -10,6 +10,7 @@
 namespace ZendTest\Http;
 
 use Zend\Http\Response;
+use Zend\Http\Headers;
```
Comment (webimpress) [inline test/ResponseTest.php]: Please have all imports in alphabetical order.

**Comment by domoran** [comment]
Fixed formatting comments.

**Comment by domoran** [comment]
Hmm ... the test seems flaky, failed under PHP 7 for the last build ... I cannot think of a way to make the test more stable except by increasing its runtime.  (More data, increase ratio) Any ideas?

**Comment by Ocramius** [comment]
@domoran flakiness can probably be avoided by repeating the test, but indeed, it's a time-based test, so it will always be a risky one. Make sure the first iteration is removed from measurement to avoid autolading and other possible performance issues there.

**Comment by Ocramius** [comment]
Sorry for the previous commits - previous review from me was never submitted.

File: test/ResponseTest.php [hunk 12]
```diff
@@ -176,6 +177,60 @@ public function testChunkedResponseCaseInsensitiveZF5438()
...
+     */
+    private function getTimeForGetBody(Response $response)
+    {
+        $timeStart = microtime(true);
+        $response->getBody();
+        return microtime(true) - $timeStart;
+    }
+
+    /**
+     * @small
```
Comment (Ocramius) [inline test/ResponseTest.php]: What's `@small` here?

Comment (Ocramius) [inline test/ResponseTest.php]: Does the test fail with the previous code? (just asking, since can't try it on my machine atm)

File: test/ResponseTest.php [hunk 13]
```diff
@@ -176,6 +177,60 @@ public function testChunkedResponseCaseInsensitiveZF5438()
         $this->assertEquals('c0cc9d44790fa2a58078059bab1902a9', md5($res->getContent()));
     }
 
+    /**
+     * @param int $chunksize the data size of the chunk to create
+     * @return string a chunk of data for embedding inside a chunked response
+     */
+    private function makeChunk($chunksize)
+    {
+        $chunkdata = str_repeat('W', $chunksize);
```
Comment (Ocramius) [inline test/ResponseTest.php]: Can be inlined

**Comment by domoran** [comment]
I dont get it. There must be a logic error in the test or a problem on php7. 

How can one more regexp iteration inside the decodeChunkedBody be factor [30-50(!)](https://travis-ci.org/zendframework/zend-http/jobs/199619906) slower than the baseline? This happens only on travis-ci and only with php 7 ... 

On my machine I cannot reproduce. Even with php7 I get a ratio of 3-5 ... But with php5.6 there is the expected runtime ratio of 1 ... Maybe an issue with the regexp implementation on PHP 7?

Oh boy. Committing to Zend Framework is hard work.

File: test/ResponseTest.php (hunk 12 above)
Comment (domoran) [inline test/ResponseTest.php]: On my machine the test does not fail at all. Neither with php7 nor php 5.6 ... small will be removed.

**Comment by Ocramius** [comment]
> Oh boy. Committing to Zend Framework is hard work.

Nah, you just found an annoyingly hard edge case.

Did you check whether you are running with xdebug or not? Try running tests with `php -n` (will probably fail on some curl-based scenarios).

If we can help out, let us know, we can certainly test on other environments as well.

PHP 7 has indeed changed many things in the regex engine, btw.

File: test/ResponseTest.php (hunk 12 above)
Comment (Ocramius) [inline test/ResponseTest.php]: @domoran I was asking if the test fails when removing the patch to the chunk decoding block.

Comment (domoran) [inline test/ResponseTest.php]: Yes, without the path the ratios are on my machine 2000-5000 under PHP7 and about 250 on PHP5.6 ... Test runs > 60 secs.  

I can increase the ratio here to make the test pass, but there must be some other problem. I would not expect a ratio much larger than 1, after all this is only one iteration more than before.

Comment (Ocramius) [inline test/ResponseTest.php]: Alright, going up with those timings is not gonna help, so I suppose that a different approach needs to be taken.

Could you poke me on Friday? I can probably work on it while traveling.

**Comment by weierophinney** [comment]
Thanks, @domoron! We have updated the test suite recently, and those updates ensured your performance-based tests now pass reliably! Merged for next release.
